pkg/controller/cronjob/cronjob_controller.go

```diff
@@ -202,9 +202,12 @@ func removeOldestJobs(sj *batchv1beta1.CronJob, js []batchv1.Job, jc jobControlI
 	klog.V(4).Infof("Cleaning up %d/%d jobs from %s", numToDelete, len(js), nameForLog)
 
 	sort.Sort(byJobStartTime(js))
-	for i := 0; i < numToDelete; i++ {
+	for i := 0; i < numToDelete && i < len(js); i++ {
 		klog.V(4).Infof("Removing job %s from %s", js[i].Name, nameForLog)
-		deleteJob(sj, &js[i], jc, recorder)
+		if !deleteJob(sj, &js[i], jc, recorder) {
+			// the current job is not deleted. extend numToDelete to accommodate
+			numToDelete++
```

Comment:
Let's say `deleteJob` returns false when `i == 0`, how does this accommodate that? `js[0]` won't be tried again in this loop. 

Besides, because Jobs clean up is called every 10s, it's fine to ignore this and let the controller retry. `deleteJob` already emits warnings for Job deletion failures.

Reply:
For #1, I didn't intend to retry js[0] in the loop - job deletion for js[0] may fail again.

For #2, the PR tries to handle the following scenario:

some old jobs fail to be deleted (possibly across Job clean ups). Due to the sort.Sort call on line 204, such jobs would be at the head of js array. The newer Jobs at the back of js may not be given chance to be deleted.

With this PR, the loop would try to delete numToDelete Jobs, not counting deletion failures.

What do you think, @janetkuo ?

Reply:
I see. 

CronJob controller retains X latest Jobs. This change would delete new Jobs if older ones fail to get deleted. This is not the behavior the CronJob controller supports.